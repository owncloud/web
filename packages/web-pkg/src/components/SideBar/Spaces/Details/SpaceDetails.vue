<template>
  <div id="oc-space-details-sidebar">
    <div class="oc-space-details-sidebar-image oc-text-center">
      <oc-spinner v-if="loadImageTask.isRunning" />
      <div v-else-if="spaceImage" class="oc-position-relative">
        <img :src="spaceImage" alt="" class="oc-mb-s" />
      </div>
      <oc-icon
        v-else
        name="layout-grid"
        size="xxlarge"
        class="space-default-image oc-px-m oc-py-m"
      />
    </div>
    <div
      v-if="showShareIndicators && hasShares"
      class="oc-flex oc-flex-middle oc-space-details-sidebar-members oc-mb-s oc-text-small"
      style="gap: 15px"
    >
      <oc-button
        v-if="hasMemberShares"
        appearance="raw"
        :aria-label="openSharesPanelMembersHint"
        @click="expandSharesPanel"
      >
        <oc-icon name="group" />
      </oc-button>
      <oc-button
        v-if="hasLinkShares"
        appearance="raw"
        :aria-label="openSharesPanelLinkHint"
        @click="expandSharesPanel"
      >
        <oc-icon name="link" />
      </oc-button>
      <p v-text="shareLabel" />
      <oc-button
        appearance="raw"
        :aria-label="openSharesPanelHint"
        size="small"
        @click="expandSharesPanel"
      >
        <span class="oc-text-small" v-text="$gettext('Show')" />
      </oc-button>
    </div>
    <table class="details-table oc-width-1-1" :aria-label="detailsTableLabel">
      <col class="oc-width-1-3" />
      <col class="oc-width-2-3" />
      <tr>
        <th scope="col" class="oc-pr-s oc-font-semibold" v-text="$gettext('Last activity')" />
        <td v-text="lastModifiedDate" />
      </tr>
      <tr v-if="resource.description">
        <th scope="col" class="oc-pr-s oc-font-semibold" v-text="$gettext('Subtitle')" />
        <td v-text="resource.description" />
      </tr>
      <tr>
        <th scope="col" class="oc-pr-s oc-font-semibold" v-text="$gettext('Manager')" />
        <td>
          <span v-text="ownerUsernames" />
        </td>
      </tr>
      <tr v-if="!resource.disabled">
        <th scope="col" class="oc-pr-s oc-font-semibold" v-text="$gettext('Quota')" />
        <td>
          <space-quota :space-quota="resource.spaceQuota" />
        </td>
      </tr>
      <tr v-if="showSize" data-testid="sizeInfo">
        <th scope="col" class="oc-pr-s oc-font-semibold" v-text="$gettext('Size')" />
        <td v-text="size" />
      </tr>
      <web-dav-details v-if="showWebDavDetails" :space="resource" />
      <portal-target
        name="app.files.sidebar.space.details.table"
        :slot-props="{ space: resource, resource }"
        :multiple="true"
      />
    </table>
  </div>
</template>
<script lang="ts">
import { storeToRefs } from 'pinia'
import { defineComponent, inject, ref, Ref, computed, unref } from 'vue'
import { useTask } from 'vue-concurrency'
import { getRelativeSpecialFolderSpacePath, SpaceResource } from '@ownclouders/web-client'
import {
  usePreviewService,
  useClientService,
  useUserStore,
  useSpacesStore,
  useSharesStore,
  useResourcesStore,
  useResourceContents,
  useRouter
} from '../../../../composables'
import SpaceQuota from '../../../SpaceQuota.vue'
import WebDavDetails from '../../WebDavDetails.vue'
import { formatDateFromISO, formatFileSize } from '../../../../helpers'
import { eventBus } from '../../../../services/eventBus'
import { SideBarEventTopics } from '../../../../composables'
import { ImageDimension } from '../../../../constants'
import { ProcessorType } from '../../../../services'
import { isLocationSpacesActive } from '../../../../router'
import { useGettext } from 'vue3-gettext'

export default defineComponent({
  name: 'SpaceDetails',
  components: { SpaceQuota, WebDavDetails },
  props: {
    showSpaceImage: {
      type: Boolean,
      required: false,
      default: true
    },
    showShareIndicators: {
      type: Boolean,
      required: false,
      default: true
    }
  },
  setup(props) {
    const userStore = useUserStore()
    const previewService = usePreviewService()
    const clientService = useClientService()
    const spacesStore = useSpacesStore()
    const resourcesStore = useResourcesStore()
    const { resourceContentsText } = useResourceContents({ showSizeInformation: false })
    const router = useRouter()
    const { current: currentLanguage } = useGettext()

    const sharesStore = useSharesStore()
    const { spaceMembers } = storeToRefs(spacesStore)

    const resource = inject<Ref<SpaceResource>>('resource')
    const spaceImage = ref('')

    const { user } = storeToRefs(userStore)

    const loadImageTask = useTask(function* (signal, ref) {
      if (!ref.resource?.spaceImageData || !props.showSpaceImage) {
        spaceImage.value = undefined
        return
      }

      const imageResource = yield clientService.webdav.getFileInfo(unref(resource), {
        path: getRelativeSpecialFolderSpacePath(unref(resource), 'image')
      })

      spaceImage.value = yield previewService.loadPreview({
        space: unref(resource),
        resource: imageResource,
        dimensions: ImageDimension.Tile,
        processor: ProcessorType.enum.fit
      })
    })

    const linkShareCount = computed(() => sharesStore.linkShares.length)
    const showWebDavDetails = computed(() => resourcesStore.areWebDavDetailsShown)
    const showSize = computed(() => {
      return !isLocationSpacesActive(router, 'files-spaces-projects')
    })
    const size = computed(() => {
      return `${formatFileSize(unref(resource).size, currentLanguage)}, ${unref(
        resourceContentsText
      )}`
    })

    return {
      loadImageTask,
      spaceImage,
      resource,
      linkShareCount,
      showWebDavDetails,
      user,
      spaceMembers,
      resourceContentsText,
      showSize,
      size
    }
  },
  computed: {
    hasShares() {
      return this.hasMemberShares || this.hasLinkShares
    },
    shareLabel() {
      if (this.hasMemberShares && !this.hasLinkShares) {
        return this.memberShareLabel
      }
      if (!this.hasMemberShares && this.hasLinkShares) {
        return this.linkShareLabel
      }

      switch (this.memberShareCount) {
        case 1:
          return this.$ngettext(
            'This space has one member and %{linkShareCount} link.',
            'This space has one member and %{linkShareCount} links.',
            this.linkShareCount,
            { linkShareCount: this.linkShareCount.toString() }
          )
        default:
          if (this.linkShareCount === 1) {
            return this.$gettext('This space has %{memberShareCount} members and one link.', {
              memberShareCount: this.memberShareCount.toString()
            })
          }
          return this.$gettext(
            'This space has %{memberShareCount} members and %{linkShareCount} links.',
            {
              memberShareCount: this.memberShareCount.toString(),
              linkShareCount: this.linkShareCount.toString()
            }
          )
      }
    },
    openSharesPanelHint() {
      return this.$gettext('Open share panel')
    },
    openSharesPanelLinkHint() {
      return this.$gettext('Open link list in share panel')
    },
    openSharesPanelMembersHint() {
      return this.$gettext('Open member list in share panel')
    },
    detailsTableLabel() {
      return this.$gettext('Overview of the information about the selected space')
    },
    lastModifiedDate() {
      return formatDateFromISO(this.resource.mdate, this.$language.current)
    },
    ownerUsernames() {
      return this.resource.spaceRoles.manager
        .map((share) => {
          if (share.id === this.user?.id) {
            return this.$gettext('%{displayName} (me)', { displayName: share.displayName })
          }
          return share.displayName
        })
        .join(', ')
    },
    hasMemberShares() {
      return this.memberShareCount > 0
    },
    hasLinkShares() {
      return this.linkShareCount > 0
    },
    memberShareCount() {
      return this.spaceMembers.length
    },
    memberShareLabel() {
      return this.$ngettext(
        'This space has %{memberShareCount} member.',
        'This space has %{memberShareCount} members.',
        this.memberShareCount,
        { memberShareCount: this.memberShareCount.toString() }
      )
    },
    linkShareLabel() {
      return this.$ngettext(
        '%{linkShareCount} link giving access.',
        '%{linkShareCount} links giving access.',
        this.linkShareCount,
        { linkShareCount: this.linkShareCount.toString() }
      )
    }
  },
  watch: {
    'resource.spaceImageData': {
      handler() {
        this.loadImageTask.perform(this)
      },
      deep: true
    }
  },
  mounted() {
    this.loadImageTask.perform(this)
  },
  methods: {
    expandSharesPanel() {
      eventBus.publish(SideBarEventTopics.setActivePanel, 'space-share')
    }
  }
})
</script>
<style lang="scss" scoped>
.oc-space-details-sidebar {
  &-image img {
    max-height: 150px;
    object-fit: cover;
    width: 100%;
  }
}

.details-table {
  text-align: left;
  table-layout: fixed;

  tr {
    height: 1.5rem;
  }
}
</style>
