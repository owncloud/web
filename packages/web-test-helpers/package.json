{
  "name": "@ownclouders/web-test-helpers",
  "version": "11.3.2",
  "description": "ownCloud web test helpers",
  "license": "AGPL-3.0",
  "private": false,
  "type": "module",
  "main": "src/index.ts",
  "author": "ownCloud GmbH <devops@owncloud.com>",
  "homepage": "https://github.com/owncloud/web/tree/master/packages/web-test-helpers",
  "repository": {
    "type": "git",
    "url": "https://github.com/owncloud/web",
    "directory": "packages/web-test-helpers"
  },
  "files": [
    "dist/"
  ],
  "publishConfig": {
    "directory": "package",
    "linkDirectory": false,
    "exports": {
      ".": {
        "import": "./dist/web-test-helpers.es.js",
        "require": "./dist/web-test-helpers.umd.cjs",
        "types": "./dist/src/index.d.ts"
      }
    }
  },
  "scripts": {
    "vite": "vite",
    "prepublishOnly": "rm -rf ./package && clean-publish && find package && cat package/package.json",
    "postpublish": "rm -rf ./package"
  },
  "peerDependencies": {
    "@ownclouders/web-pkg": "workspace:^",
    "@vue/test-utils": "^2.4.6",
    "vue": "^3.5.10"
  },
  "devDependencies": {
    "@vitejs/plugin-vue": "5.1.4",
    "clean-publish": "^5.0.0",
    "vite-plugin-dts": "^4.2.3",
    "vite": "^5.4.8"
  },
  "dependencies": {
    "@casl/ability": "^6.7.1",
    "@casl/vue": "^2.2.2",
    "@ownclouders/design-system": "workspace:^",
    "@ownclouders/web-client": "workspace:^",
    "@pinia/testing": "^0.1.3",
    "axios": "1.7.7",
    "pinia": "2.2.6",
    "vitest-mock-extended": "2.0.2",
    "vue-router": "4.2.5",
    "vue3-gettext": "2.4.0"
  }
}
