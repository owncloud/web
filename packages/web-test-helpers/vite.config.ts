import { resolve } from 'path'
import { defineConfig } from 'vite'
import dts from 'vite-plugin-dts'
import pkg from './package.json' assert { type: 'json' }
import vue from '@vitejs/plugin-vue'

export default defineConfig({
  build: {
    lib: {
      entry: resolve(__dirname, 'src/index.ts'),
      name: 'web-test-helpers',
      fileName: (format) => `web-test-helpers.${format}.js`
    },
    rollupOptions: {
      external: [...Object.keys(pkg.dependencies), ...Object.keys(pkg.peerDependencies)]
    }
  },
  plugins: [vue(), dts({ include: ['src'], outDir: 'dist/types', insertTypesEntry: true })]
})
