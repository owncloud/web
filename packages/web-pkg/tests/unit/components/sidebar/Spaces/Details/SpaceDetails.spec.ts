import SpaceDetails from '../../../../../../src/components/SideBar/Spaces/Details/SpaceDetails.vue'
import {
  CollaboratorShare,
  ShareRole,
  GraphShareRoleIdMap,
  SpaceResource,
  Resource
} from '@ownclouders/web-client'
import { mock } from 'vitest-mock-extended'
import { defaultComponentMocks, defaultPlugins, shallowMount } from 'web-test-helpers'
import { RouteLocation } from 'vue-router'
import { User } from '@ownclouders/web-client/graph/generated'

const spaceMock = {
  type: 'space',
  name: ' space',
  id: '1',
  mdate: 'Wed, 21 Oct 2015 07:28:00 GMT',
  spaceRoles: {
    manager: [{ id: '1', displayName: 'alice' }],
    editor: [],
    viewer: []
  },
  spaceQuota: {
    used: 100,
    total: 1000
  }
} as unknown as SpaceResource

const spaceShare = {
  id: '1',
  sharedWith: {
    id: 'Alice',
    displayName: 'alice'
  },
  role: mock<ShareRole>({ id: GraphShareRoleIdMap.SpaceManager })
} as CollaboratorShare

const selectors = {
  spaceDefaultImage: '.space-default-image',
  spaceMembers: '.oc-space-details-sidebar-members'
}

describe('Details SideBar Panel', () => {
  it('displays the details side panel', () => {
    const { wrapper } = createWrapper()
    expect(wrapper.html()).toMatchSnapshot()
  })
  it('does render the space default image if "showSpaceImage" is false', () => {
    const { wrapper } = createWrapper({ props: { showSpaceImage: false } })
    expect(wrapper.find(selectors.spaceDefaultImage).exists()).toBeTruthy()
  })
  it('does not render share indicators if "showShareIndicators" is false', () => {
    const { wrapper } = createWrapper({
      spaceResource: spaceMock,
      props: { showShareIndicators: false }
    })
    expect(wrapper.find(selectors.spaceMembers).exists()).toBeFalsy()
  })
  it('does not render share indicators if space is disabled', () => {
    const { wrapper } = createWrapper({
      spaceResource: { ...spaceMock, disabled: true },
      props: { showShareIndicators: true }
    })
    expect(wrapper.find(selectors.spaceMembers).exists()).toBeFalsy()
  })
})

function createWrapper({ spaceResource = spaceMock, props = {} } = {}) {
  const mocks = {
    ...defaultComponentMocks({
      currentRoute: mock<RouteLocation>({ name: 'files-spaces-generic' })
    })
  }

  return {
    wrapper: shallowMount(SpaceDetails, {
      props: { ...props },
      global: {
        plugins: [
          ...defaultPlugins({
            piniaOptions: {
              userState: { user: { id: '1', onPremisesSamAccountName: 'marie' } as User },
              spacesState: { spaceMembers: [spaceShare] },
              sharesState: { collaboratorShares: [spaceShare] },
              resourcesStore: { resources: [mock<Resource>({ name: 'file1', type: 'file' })] }
            }
          })
        ],
        mocks,
        provide: { resource: spaceResource, ...mocks }
      }
    })
  }
}
