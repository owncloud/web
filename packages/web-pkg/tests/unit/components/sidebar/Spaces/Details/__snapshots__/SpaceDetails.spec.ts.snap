// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Details SideBar Panel > displays the details side panel 1`] = `
"<div data-v-c7ea2f9c="" id="oc-space-details-sidebar">
  <div data-v-c7ea2f9c="" class="oc-space-details-sidebar-image oc-text-center">
    <oc-icon-stub data-v-c7ea2f9c="" name="layout-grid" filltype="fill" accessiblelabel="" type="span" size="xxlarge" variation="passive" color="" class="space-default-image oc-px-m oc-py-m"></oc-icon-stub>
  </div>
  <div data-v-c7ea2f9c="" class="oc-flex oc-flex-middle oc-space-details-sidebar-members oc-mb-s oc-text-small" style="gap: 15px;">
    <oc-button-stub data-v-c7ea2f9c="" type="button" disabled="false" size="medium" arialabel="Open member list in share panel" submit="button" variation="passive" appearance="raw" justifycontent="center" gapsize="medium" showspinner="false"></oc-button-stub>
    <!--v-if-->
    <p data-v-c7ea2f9c="">This space has 1 member.</p>
    <oc-button-stub data-v-c7ea2f9c="" type="button" disabled="false" size="small" arialabel="Open share panel" submit="button" variation="passive" appearance="raw" justifycontent="center" gapsize="medium" showspinner="false"></oc-button-stub>
  </div>
  <table data-v-c7ea2f9c="" class="details-table oc-width-1-1" aria-label="Overview of the information about the selected space">
    <col data-v-c7ea2f9c="" class="oc-width-1-3">
    <col data-v-c7ea2f9c="" class="oc-width-2-3">
    <tr data-v-c7ea2f9c="">
      <th data-v-c7ea2f9c="" scope="col" class="oc-pr-s oc-font-semibold">Last activity</th>
      <td data-v-c7ea2f9c="">Invalid DateTime</td>
    </tr>
    <!--v-if-->
    <tr data-v-c7ea2f9c="">
      <th data-v-c7ea2f9c="" scope="col" class="oc-pr-s oc-font-semibold">Manager</th>
      <td data-v-c7ea2f9c=""><span data-v-c7ea2f9c="">alice (me)</span></td>
    </tr>
    <tr data-v-c7ea2f9c="">
      <th data-v-c7ea2f9c="" scope="col" class="oc-pr-s oc-font-semibold">Quota</th>
      <td data-v-c7ea2f9c="">
        <space-quota-stub data-v-c7ea2f9c="" spacequota="[object Object]"></space-quota-stub>
      </td>
    </tr>
    <tr data-v-c7ea2f9c="" data-testid="sizeInfo">
      <th data-v-c7ea2f9c="" scope="col" class="oc-pr-s oc-font-semibold">Size</th>
      <td data-v-c7ea2f9c="">?, 1 item in total (1 file, 0 folders)</td>
    </tr>
    <!--v-if-->
    <portal-target data-v-c7ea2f9c="" name="app.files.sidebar.space.details.table" slot-props="[object Object]" multiple="true"></portal-target>
  </table>
</div>"
`;
